### PR TITLE
Changes to enable vectorization in ParallelAccelerator.

### PR DESCRIPTION
**Pull request by DrTodd13**
I started from a clean master repo including the recently merged ascontiguousarray PR and then cherry-picked my commits into the branch so things look much cleaner now.  All the TestParfors tests work.

### CODE REVIEW

**Comment by DrTodd13** [comment]
One potential issue is that I was only able to easily do alias analysis for a single parfor during lowering to determine whether to add the noalias flag or not.  This fixed the issue with one of the tests that we were seeing but I'm not sure if it is sufficient in general or not.  Why couldn't aliases be established outside the parfor that wouldn't be caught with this method?

Another issue we need to solve before we merge is how to handle prange with negative starting index.  Right now the returned schedule is always uintp so it would definitely fail now.

**Comment by DrTodd13** [comment]
Also, blackscholes without the transcendentals still seems to vectorize without the ascontiguousarray stuff.

File: numba/npyufunc/parfor.py [hunk 1]
```diff
@@ -592,7 +633,16 @@ def _create_gufunc_for_parfor_body(
     if config.DEBUG_ARRAY_OPT:
         print("gufunc_ir last dump")
         gufunc_ir.dump()
+        print("flags", flags)
+        print("typemap", typemap)
 
+    old_alias = flags.noalias
```
[the remark this answers is not in the document]
Comment (DrTodd13) [reply numba/npyufunc/parfor.py]: Do you think it is okay to add error_model=numpy?  We have to add noalias.
I will remove the fastmath part and propagate flags for now in the parfor.

File: numba/typing/builtins.py [hunk 2]
```diff
@@ -953,7 +953,7 @@ def typeof_index(val, c):
 @type_callable(IndexValue)
 def type_index_value(context):
     def typer(ind, mval):
-        if ind == types.intp:
+        if ind == types.intp or ind == types.uintp:
```
[the remark this answers is not in the document]
Comment (DrTodd13) [reply numba/typing/builtins.py]: You think there needs to be an unsigned version of IndexValueModel?  To be honest, this was one of those "find a bug and fix it without understanding it things."  I know if the type was unsigned that it wasn't recognized so I added the unintp.  Depending on what IndexValueModel is used for may decide whether this is sufficiently consistent.  Thoughts?

File: numba/npyufunc/parfor.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (stuartarchibald) [reply numba/npyufunc/parfor.py]: I guess the consequence of not using the NumPy error model is code with more branches that won't vectorize so well/performance may be worse. With regards to setting `error_model='numpy'` as default, the use of `gufunc` and related machinery is an implementation detail of parfors and as a result I'm unsure that setting this flag as noted is desirable. Users will be encountering the parfors transforms through `parallel=True` (and possibly `prange`), I think it is therefore likely that they will have `njit(parallel=True)` decorated their function that uses Python+NumPy expecting it to work approximately the same way as without the decoration. For this reason, even though setting the error model to `'numpy'` potentially allows for more aggressive optimisation, I think it still needs to be opt-in. Perhaps we could raise a performance warning if division is encountered in the kernel and the error model is set to `python`? A similar warning could also be raised if a user sets `fastmath` but retains the `python` error model? TBH I think it's best to open the flag combination/performance/user expectation trade off up to discussion... @seibert @sklam @ehsantn any views on this?

I think perhaps the following can be done without further input?
 * Remove the `fastmath` hardcoding and do the propagation of parfors `flags` as noted.
 * Leave `noalias` as is as its necessity is determined by analysis.

As an aside, I think having `fastmath` and the NumPy error model could result in the violation of the expectation set by the `nnan` and `ninf` flags implied by `fast` in the case of a div-by-zero. I anticipate in the event of such a violation there would, however, be little surprise as a user setting `fastmath` would likely be aware of the unsafe FP transforms requested.

**Comment by stuartarchibald** [comment]
I'm fairly sure equivalence analysis is failing to pick up that `prange`s with identical negative inferred-by-variable range are equivalent:

```python
import numpy as np
from numba import njit, prange

def test1(n):
    # fuse adjacent +ve pranges
    acc = acc2 = 0
    for i in prange(n):
        acc += i
    for i in prange(n):
        acc2 += i
    return acc + acc2

def test2(n):
    # fuse +ve prange with function of same dimension and parallel semantic
    acc = 0
    for i in prange(n):
        acc += i
    f = np.zeros(n)
    return acc + f[0]

def test3(n):
    # fuse adjacent -ve const pranges
    acc = acc2 = 0
    for i in prange(-4, 0):
        acc += i
    for i in prange(-4, 0):
        acc2 += i
    return acc + acc2

def test4(n):
    # fuse adjacent -ve pranges by var..... fails
    acc = acc2 = 0
    for i in prange(-n, 0):
        acc += i
    for i in prange(-n, 0):
        acc2 += i
    return acc + acc2


def tester(func):
    print("Testing %s" % func.__name__)
    njit(parallel=True)(func)(5)
    print("\n")

tests = [test1, test2, test3, test4]

[tester(x) for x in tests]
```
gives:
```
Testing test1
Parallel for-loop #0 is produced from pattern '('prange', 'user')' at test_argmax.py (7)
Parallel for-loop #1 is produced from pattern '('prange', 'user')' at test_argmax.py (9)
Parallel for-loop #1 is fused into for-loop #0.
After fusion, function test1 has 1 parallel for-loop(s) #{0}.


Testing test2
Parallel for-loop #2 is produced from pattern 'zeros function' at test_argmax.py (18)
Parallel for-loop #3 is produced from pattern '('prange', 'user')' at test_argmax.py (16)
Parallel for-loop #2 is fused into for-loop #3.
After fusion, function test2 has 1 parallel for-loop(s) #{3}.


Testing test3
Parallel for-loop #4 is produced from pattern '('prange', 'user')' at test_argmax.py (24)
Parallel for-loop #5 is produced from pattern '('prange', 'user')' at test_argmax.py (26)
Parallel for-loop #5 is fused into for-loop #4.
After fusion, function test3 has 1 parallel for-loop(s) #{4}.


Testing test4
Parallel for-loop #6 is produced from pattern '('prange', 'user')' at test_argmax.py (33)
Parallel for-loop #7 is produced from pattern '('prange', 'user')' at test_argmax.py (35)
After fusion, function test4 has 2 parallel for-loop(s) #{6, 7}.
```
I guess this definition is the problem? https://github.com/numba/numba/blob/master/numba/array_analysis.py#L1232-L1238

Would you prefer to address this in a different patchset?

File: numba/npyufunc/parfor.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (seibert) [reply numba/npyufunc/parfor.py]: yeah, the flag question is important.  I think perhaps we could do the following:
* `fastmath=True` implies `error_model="numpy"` for the reasons you describe.
* `parallel=True` does not imply `fastmath=True`, but does print a warning (once) suggesting the user try `fastmath=True`.  We would need an entry in the FAQ explaining how to disable the warning.

File: numba/typing/builtins.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (stuartarchibald) [reply numba/typing/builtins.py]: I *think* it's ok at present. Having a signed and unsigned `IndexValue` implementation would mean having to write code to unify them etc (also conveniently `intp/uintp` are the same width). I think that the looser definition actually allows parfors to pick the type at a later, more profitable, point in the pipeline. It was tricky looking at this and I would be delighted to be corrected if on the wrong track :)

File: numba/tests/test_parfors.py [hunk 3]
```diff
@@ -1156,7 +1156,7 @@ def test_impl():
             a = 2
             b = 3
             A = np.empty(4)
-            for i in range(-2, 4):
+            for i in range(-2, 2):
```
Comment (stuartarchibald) [inline numba/tests/test_parfors.py]: ah yes, well spotted, thanks for correcting it :)

**Comment by ehsantn** [comment]
@stuartarchibald The definitions you mentioned just says if `m = +n`, m and n are equivalent. If `m = -n`, then they are not equivalent.

The fusion test case points to a different problem: we don't have common subexpression elimination (CSE) or similar expression analysis. Therefore, if `$1 = -n` and `$2 = -n`, we don't know $1 and $2 are equivalent. We can only infer things from a single statement.

**Comment by stuartarchibald** [comment]
@ehsantn thanks, the compilation path simply just seemed to point to breakdown there, but as noted additional work is needed. There are two different `-n` assignments in the case presented, I think CSE may be generally useful if that's the best way around this.

**Comment by stuartarchibald** [comment]
I've added a patch here: https://github.com/stuartarchibald/numba/commit/026eb6902a3f1d6b5da97e37acdbc946b9324e33 which does a loose check on the application of `fastmath` in that it essentially checks that if the `fastmath` flags are added then instructions that can be `fast` are found in the IR. Any thoughts on this approach?

I think that to check that vectorization is working is going to be quite hard because even in a trivial case, like a stride-1 reduction loop, the instruction selection is going to be architecture dependent. We could therefore do one of two things.
 * Expose the `LLVM` vectorisation reports.
 * Use https://github.com/numba/numba/pull/2515 to target some known feature set and ensure that in the trivial case the instruction sequence generated is what is expected with/without `fastmath`.

**Comment by stuartarchibald** [comment]
@DrTodd13 
Looking at the failing test, the message is:
```
FAIL: test_array_debug_opt_stats (numba.tests.test_debug.TestParforsDebug)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/build/numba/numba/numba/tests/test_debug.py", line 280, in test_array_debug_opt_stats
    self.assertIn(to_match, trials)
AssertionError: "Parallel for-loop #0 is produced from pattern 'ones function' at" not found in "Parallel for-loop #6 is produced from pattern 'ones function' at /home/travis/build/numba/numba/numba/tests/test_debug.py (43)"
```

I think this is probably because the `Parfor` class has a class variable `id_counter`, which means that diagnostics are produced based on the shared state of `Parfor`s transforms instances of which `6` have evidently appeared in the unit tests at this point.  This also highlights that given the init sequence for the `Parfor` class that the `Parfor` pipeline itself is not threadsafe.

PR: https://github.com/IntelLabs/numba/pull/57 attempts to fix the test along with a couple of other minor issues.

Outstanding items are:
 1. Contiguousness needs work.
 2. Slicing analysis needs work.
 3. The `error_model` propagation needs addressing (and work in general, there is a lot of state twiddling in the code base).
 4. CSE pass or similar to accommodate https://github.com/numba/numba/pull/2727#issuecomment-366418807

I'll open tickets for the above once this is merged.

Should the above PR be merged and tests pass I think this is good to go, thanks again for the patches.

**Comment by DrTodd13** [comment]
Yes, let's merge it.

**Comment by stuartarchibald** [comment]
Tests all pass now. Thanks for all the effort in writing and resolving the issues in this. Merging!

**Pull request merged**